### PR TITLE
Issue/4972 pending drafts backoff refactor

### LINKED ISSUE #4972: Notifications: enhance pending drafts notifications texts and behavior
Having one 24h reminder is a thing but having the same notification coming back every day is showing up to not be useful but rather annoying.

Have +1 day, +1 week, +1 month reminders, with different messages randomly chosen, that could even be fun:

- 1 day:

```
You drafted “post title” yesterday. Don’t forget to publish it!
Did you know that “post title” is still a draft? Publish away!
```

- 1 week:

```
Your draft, “post title” awaits you — be sure to publish it!
“Post title” remains a draft. Remember to publish it!
```

- 1 month

```You drafted “post title” X days ago. Your audience awaits it!
Don’t leave it hanging! “Post title” is waiting to be published.
```

### PR DESCRIPTION
**Pull request by mzorz**
Fixes #4972

This PR goes with the architecture idea proposed by @maxme and thus ditches the intertwined polling process we had to check for pending drafts.

As soon as you exit a post draft by pressing back without publishing it, this PR creates a +1 day, +1 week, +1 month reminders, with different messages randomly chosen as follows:

- 1 day:

```
You drafted “post title” yesterday. Don’t forget to publish it!
Did you know that “post title” is still a draft? Publish away!
```

- 1 week:

```
Your draft, “post title” awaits you — be sure to publish it!
“Post title” remains a draft. Remember to publish it!
```

- 1 month

```You drafted “post title” X days ago. Your audience awaits it!
Don’t leave it hanging! “Post title” is waiting to be published.
```

To test:
1. Login to the app, and go to Blog Posts
2. Start a new draft by pressing the + floating button
3. write something
4. press back
5. change the device date setting to +1 day, and a new notification should appear. Check that it is one of the +1 type cited above. If you tap EDIT it take you to the draft. Tapping IGNORE will dismiss it.
6. change the device date to +1 week. If the previous notification wasn't dismissed, you'll notice the text for it will have changed to one of the types +1 week cited above. If you dismissed it, a new notification appears, with either 1 of the 2 texts cited above.
7. change the device date to +1 month, you should see a new notification (or if it was there, the existing notification text changes) to the month-version text as cited above.

cc @maxme for review, and cc'ing @daniloercoli as well

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java [hunk 1]
```diff
@@ -859,7 +868,71 @@ public boolean isNewPost() {
...
+        return alarmIntentOneDay;
+    }
+
+    private PendingIntent getOneWeekAlarmIntent(Context context, Intent notifPendingDraftReceiverIntent) {
+        PendingIntent alarmIntentOneWeek = PendingIntent.getBroadcast(context, BROADCAST_BASE_REQUEST_CODE + 1, notifPendingDraftReceiverIntent, PendingIntent.FLAG_UPDATE_CURRENT);
+        return alarmIntentOneWeek;
+    }
+
+    private PendingIntent getOneMonthAlarmIntent(Context context, Intent notifPendingDraftReceiverIntent) {
+        PendingIntent alarmIntentOneMonth = PendingIntent.getBroadcast(context, BROADCAST_BASE_REQUEST_CODE + 1, notifPendingDraftReceiverIntent, PendingIntent.FLAG_UPDATE_CURRENT);
```
Comment (daniloercoli) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: `BROADCAST_BASE_REQUEST_CODE + 1` it's the same value of the `Week` branch. Is it ok?

**Comment by daniloercoli** [comment]
It seems that i'm getting only 1 notification for a draft post on a site of mine.  The app correctly builds and shows the first reminder, the "one day" notification, but if I dismiss it i'm not getting the 2d one, the "week" notification reminder.

- I created a new local draft, and then changed the device date setting to +1 day.
- A new notification appeared. Tapped IGNORE.
- Changed the device date to +1 week, and there is not a new notification.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java (hunk 1 above)
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: fixed in cd24d6c

**Comment by mzorz** [comment]
Gotcha - it was the bad number (`+1` instead of `+2`) for the request code in the BroadcastReceiver PendingIntent that was making the month/week alarms overlap. Nice catch! Also brought some more considerations when deleting the post in PostListFragment and the notification is present, so the right notification is dimissed (0747a18). 
Ready for another round.

**Comment by daniloercoli** [comment]
There are probably issues when multiple local drafts are available on the site.
I noticed that when I have 2 local drafts the `NotificationsPendingDraftsReceiver` is called twice with the same `local post ID`, resulting in a notification that doesn't use the plural form, and also points to a single Post.

File: WordPress/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -341,8 +341,13 @@
     <string name="comment_q_action_replying">Replying…</string>
 
     <!-- pending draft local notifications -->
-    <string name="pending_draft_one_generic">You drafted \'%s\' but never published it.</string>
-    <string name="pending_draft_one">You drafted \'%1$s\' %2$d days ago but never published it.</string>
+    <string name="pending_draft_one_day_1">You drafted \'%1$s\' yesterday. Don\'t forget to publish it!</string>
+    <string name="pending_draft_one_day_2">Did you know that \'%1$s\' is still a draft? Publish away!</string>
+    <string name="pending_draft_one_week_1">Your draft, \'%1$s\' awaits you - be sure to publish it!</string>
+    <string name="pending_draft_one_week_2">\'%1$s\' remains a draft. Remember to publish it!</string>
```
Comment (daniloercoli) [inline WordPress/src/main/res/values/strings.xml]: Isn't `Remember to publish them!` the correct wording for this string?

Comment (daniloercoli) [inline WordPress/src/main/res/values/strings.xml]: Sorry, I thought we were using two strings, one for the singular and one for the plural case. Instead it seems we're using 2 strings, randomly picked at run-time.

**Comment by mzorz** [comment]
> There are probably issues when multiple local drafts are available on the site.
I noticed that when I have 2 local drafts the NotificationsPendingDraftsReceiver is called twice with the same local post ID, resulting in a notification that doesn't use the plural form, and also points to a single Post.

It's true its being called twice - just a caveat the feature is not particularly designed for the "multiple local drafts" case, that means it only triggers the alarms for the _last_ post you edit.

I think the following could be food for another issue / PR:
- We could make notifications for each local draft but that would clutter if many local drafts were available
- We could try and have the plural forms as we used to have in the previous (current released) version.

**Comment by daniloercoli** [comment]
Yeah, sorry I was confused about that. I thought we were using plurals since there were 2 strings for each type of notification. Instead one of the two is picked at run-time.

File: WordPress/src/main/java/org/wordpress/android/push/NotificationsProcessingService.java [hunk 3]
```diff
@@ -178,24 +169,22 @@ public void process() {
...
+                    AnalyticsTracker.track(AnalyticsTracker.Stat.NOTIFICATION_PENDING_DRAFTS_IGNORED);
                     return;
                 }
 
                 // check special cases for pending draft notifications - dismiss
                 if (mActionType.equals(ARG_ACTION_DRAFT_PENDING_DISMISS)) {
                     //dismiss notif
-                    NativeNotificationsUtils.dismissNotification(PENDING_DRAFTS_NOTIFICATION_ID, mContext);
+                    NativeNotificationsUtils.dismissNotification(
+                            NotificationsPendingDraftsReceiver.PENDING_DRAFTS_NOTIFICATION_ID, mContext);
```
Comment (daniloercoli) [inline WordPress/src/main/java/org/wordpress/android/push/NotificationsProcessingService.java]: Should not this be `NotificationsPendingDraftsReceiver.makePendingDraftNotificationId(postId)` ?

Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/push/NotificationsProcessingService.java]: Actually when the user dismisses the notification (swipe off) there is no need to cancel such notification, so I ditched that code (which in any case was wrong as you correctly pointed out) in commit ee98223

**Comment by daniloercoli** [comment]
Hey Mario, it seems that when the device is restarted the alarms already set will be lost.
We may want to reschedule them at boot time (which also requires the android.permission.RECEIVE_BOOT_COMPLETED permission). See http://stackoverflow.com/a/12034402

**Comment by mzorz** [comment]
> Hey Mario, it seems that when the device is restarted the alarms already set will be lost.
We may want to reschedule them at boot time 

That's right, more info here (for the record) https://developer.android.com/training/scheduling/alarms.html#boot

I saw it but wasn't sure if we want to add another permission just for this. What do you think?

**Comment by daniloercoli** [comment]
Not sure how many people use to shut their devices off at night. I am doing it. I know people that does that, and people doesn't. Hard to say a percentage.
For all of those person that turn devices off, this feature is useless if we don't reschedule alarms at boot time. (Also remember that one month is a long life span for restarting. Maybe people that don't turn it off at night, restart the device within one month).

Anyway, wait for @maxme on this. I think if we want to use this new way (AlarmManager) of scheduling pending draft notifications, we should reschedule alarms at booting time. Otherwise don't do it this way, since it will affect only a small percentage of our usersbase.

**Comment by mzorz** [comment]
> Otherwise don't do it this way, since it will affect only a small percentage of our usersbase.
That's a fair point! Will implement and get back to you 👍

**Comment by daniloercoli** [comment]
> That's a fair point! Will implement and get back to you 👍

Probably better to wait until Maxime feedback, about adding `RECEIVE_BOOT_COMPLETED` permission to the app.

**Comment by maxme** [comment]
We don't have the choice if we want to implement it correctly, so add the new permission.

**Comment by mzorz** [comment]
addressed `RECEIVE_BOOT_COMPLETED` in c829ffd

Logcat after turning phone back on:

```
01-25 14:48:47.493 2638-2638/org.wordpress.android.beta I/WordPress-NOTIFS: entering Pending Drafts Receiver from BOOT_COMPLETED
```

Lockscreen after turning phone back on:
screen shot 2017-01-25 at 4 49 26 pm

**Comment by daniloercoli** [comment]
Woke up this morning and found out duplicate notifications on my device.
screen shot 2017-01-26 at 09 58 33

Not sure if this happened since I switched back ad forth the device date yesterday. I don't think it is cause of the issue, maybe something to take into consideration.

**Comment by mzorz** [comment]
_side note: those are not dupes but rather one-day notification for 2 different posts, check the titles there ;)_

Ok checked and revisited the functionality in this (rather big) commit here: 2f1c6d7

+ multiple drafts can be managed now, and each has their own 1 day, 1 week, 1 month notifications
+ edit/ignore/tap on notification body intents are handled correctly when multiple notifications are present, each one directing the user to the correct draft post
+ restarting the device now resets the schedulers rather than the notifications (previously we’d just trigger notifications when the device restarted, when we should have only reset the alarm schedules as it’s done now in these last commits). This way we make sure to count from whenever we are booting the device on to the future. 

Ready for another pass 👍

**Comment by daniloercoli** [comment]
> side note: those are not dupes but rather one-day notification for 2 different posts, check the titles there ;)

It's not clear to me if we can have multiple draft notifications or not. 
It seems that we have this ability from the screenshot I posted above. @mzorz Can you confirm?

**Comment by daniloercoli** [comment]
It seems that we can ship this.
Just a ping to @nbradbury -  Would you mind to take a look at this PR since it's pretty big, and want to be sure we're not missing anything.

**Comment by mzorz** [comment]
> It's not clear to me if we can have multiple draft notifications or not.
It seems that we have this ability from the screenshot I posted above. @mzorz Can you confirm?

Yes we now have multiple draft notifications - one "thread" (1 day, 1 week, 1 month) for each draft post. The screenshot above shows the notification for 1 day ago for 2 different posts, one entitled 'dd' and the other one 'aaaaaaa'

**Comment by nbradbury** [comment]
There's a typo in this notification - the comma is unnecessary.

screenshot_1485866197

**Comment by daniloercoli** [comment]
Just checked that the app isn't firing Analytics events at `RECEIVE_BOOT_COMPLETED`. 
Will fix the typo reported above, and merge this PR, if nobody is opposing to this.

**Comment by daniloercoli** [comment]
:shipit:

**Pull request merged**